### PR TITLE
[WIP] Optimization Docs

### LINKED ISSUE #328: Write optimization docs
This would include core optimizations in dask.optimize, collection specific optimizations, and rewrite rules.

### PR DESCRIPTION
**Pull request by jcrist**
Fixes #328.

Todo:
- [x] Prose example using top-level functions
- [x] Prose example using rewrite rules
- [x] Cleanup docstrings for optimization functions
- [ ] Document array specific optimizations
- [ ] Document bag specific optimizations
- [ ] Document dataframe specific optimiations

### CODE REVIEW

File: docs/source/optimize.rst [hunk 1]
```diff
@@ -1,9 +1,182 @@
...
+is the word counting function. For cheap operations like this, the
+serialization cost may be larger than the actual computation, so it may be
+faster to do the computation more than once, rather than passing the results to
+all nodes. To perform this function inlining, the ``inline_functions`` function
+can be used.
+
+.. code-block:: python
+
+    >>> from dask.optimize import inline_functions
+    >>> dsk3 = inline_functions(dsk2)
```
Comment (mrocklin) [inline docs/source/optimize.rst]: This might be missing the list of functions included below.

**Comment by mrocklin** [comment]
I'm unsure if we should include the explicit dicts or not.  Thoughts?

**Comment by mrocklin** [comment]
This is, I might add, a wonderful example.

**Comment by jcrist** [comment]
I don't think the explicit dicts should be shown, the diagrams are much clearer (in my opinion) than a printed dict output.

**Comment by mrocklin** [comment]
ok

File: docs/source/optimize.rst [hunk 2]
```diff
@@ -1,9 +1,182 @@
...
+In summary, the above operations:
+
+1. Removed tasks unncessary for the desired output using ``cull``
+2. Inlined constants using ``inline``
+3. Inlined cheap computations using ``inline_functions``, improving parallelism
+4. Fused linear tasks together to ensure they run on the same worker, using ``fuse``
+
+These are just a few of the optimizations provided in ``dask.optimize``, for
+more information see the api below.
+
```
Comment (mrocklin) [inline docs/source/optimize.rst]: Might want to add that the dask collections (array, bag, dataframe) automatically apply a well chosen set of optimizations.  Ordinary users of dask collections don't need to manage optimizations.

Comment (jcrist) [inline docs/source/optimize.rst]: That's at the top of this page. Should I repeat it down here?

Comment (mrocklin) [inline docs/source/optimize.rst]: Apparently I forgot about it, so maybe.

**Comment by jcrist** [comment]
I'm not sure if the collection specific optimizations warrant their own docs. Most of the optimizations done per collection are just applying the already described functions in a certain order, + a couple rewrite rules and tweaks. If anything, just a brief "we inline getitem and transpose calls, fuse slices, etc..." description for each collection. I'm not sure where that would go though.

**Comment by jcrist** [comment]
Other than that, this is ready for review.

**Comment by mrocklin** [comment]
Looks good.  Nice addition on rewrite rules.

**Pull request merged**